pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+	var pciAssignments []pciDevice
+
+	for _, adapter := range config.IoAdapterList {
+		logrus.Debugf("processing adapter %d %s\n", adapter.Type, adapter.Name)
+		list := aa.LookupIoBundleAny(adapter.Name)
+		// We reserved it in handleCreate so nobody could have stolen it
+		if len(list) == 0 {
+			logrus.Fatalf("IoBundle disappeared %d %s for %s\n",
+				adapter.Type, adapter.Name, domainName)
+		}
+		for _, ib := range list {
+			if ib == nil {
+				continue
+			}
+			if ib.Type == types.IoNetEth && ib.Ifname != adapter.Name {
+				// if we get here, means we have a PCI device which is in the same group
+				// as the ethernet passthrogh port, will have error if register to the kubevirt
+				logrus.Infof("Skip PCI device %s which does not match adapter %s\n", ib.Ifname, adapter.Name)
+				continue
+			}
+			if ib.UsedByUUID != config.UUIDandVersion.UUID {
+				logrus.Fatalf("IoBundle not ours %s: %d %s for %s\n",
+					ib.UsedByUUID, adapter.Type, adapter.Name,
+					domainName)
+			}
+			if ib.PciLong != "" {
+				logrus.Infof("Adding PCI device <%v>\n", ib.PciLong)
+				tap := pciDevice{pciLong: ib.PciLong, ioType: ib.Type}
+				pciAssignments = addNoDuplicatePCI(pciAssignments, tap)
+			}
+
+		}
+	}
+
+	if len(pciAssignments) > 0 {
+		// Device passthrough is a three step process in Kubevirt/Kubernetes
+		// 1) First do PCI Reserve like in kvm.go (If we are here, PCI Reserve is already done)
+		// 2) Register the  pciVendorSelector which is a PCI vendor ID and product ID tuple in the form vendor_id:product_id
+		//    with kubevirt
+		// 3) Then pass the registered names to VMI config
+		err := registerWithKV(kvClient, vmi, pciAssignments)
+		if err != nil {
+			return logError("Failed to register with Kubevirt  %v", len(pciAssignments))
+		}
+	}
+
+	vmi.Labels = make(map[string]string)
+	vmi.Labels[eveLabelKey] = domainName
+
+	// Now we have VirtualMachine Instance object, save it to config file for debug purposes
+	// and save it in context which will be used to start VM in Start() call
+	// dispName is for vmi name/handle on kubernetes
+	meta := vmiMetaData{
+		vmi:      vmi,
+		name:     kubeName,
+		domainID: int(rand.Uint32()),
+	}
+	ctx.vmiList[domainName] = &meta
+
+	vmiStr := fmt.Sprintf("%+v", vmi)
+
+	// write to config file
+	file.WriteString(vmiStr)
+
+	return nil
+}
+
+func (ctx kubevirtContext) Start(domainName string) error {
+	logrus.Debugf("Starting Kubevirt domain %s", domainName)
+
+	err := getConfig(&ctx)
+	if err != nil {
+		return err
+	}
+	kubeconfig := ctx.kubeConfig
+
+	vmis, ok := ctx.vmiList[domainName]
+	if !ok {
+		return logError("start domain %s failed to get vmlist", domainName)
+	}
+	if vmis.isPod {
+		err := StartPodContainer(kubeconfig, ctx.vmiList[domainName].pod)
+		return err
+	}
+
+	vmi := vmis.vmi
+	virtClient, err := kubeapi.GetKubevirtClientSet(kubeconfig)
+
+	if err != nil {
+		logrus.Errorf("couldn't get the kubernetes client API config: %v", err)
+		return err
+	}
+
+	// Create the VM
+	i := 5
```

Comment:
Why check 6 times?

Reply:
Actually I think we are looping forever, unless we cannot even talk to kubeapi server. I remove this code. @naiming-zededa any comments

Reply:
Actually I am wrong, we are checking for 6 times. But I see why though, we need to bail out after some tries and if VMI is not started by 60 secs then we need to bail. Otherwise watchdog will trigger on domainmgr,  @naiming-zededa can you confirm.

Reply:
Domainmgr is more patient that that.
It has a separate goroutine for each DomainConfig which handles the create, modify, and delete operations for that task.
Only the main loop in domainmgr has the watchdog hook.

You can see that in the delete/inactivate path where domainmgr can wait for minutes for a VM to halt.

So if there are no reasons for you to not wait forever here for that reason.
HOWEVER, domainmgr has its only logic for retry starting a task if it failed to boot. That uses a configurable timer (maybe 60 seconds by default??). 

So question is what type of failures do we see which made you add a retry here, and whether we can simplify this and use the existing retry in domainmgr?